pyfar/dsp/dsp.py

```diff
@@ -355,3 +355,134 @@ def _cross_fade(first, second, indices):
     result = first * window_first + second * window_second
 
     return result
+
+
+def minimum_phase(
+        signal, method='homomorphic', n_fft=None,
+        pad=False, return_magnitude_error=False):
+    """Calculate the minimum phase equivalent of a signal or filter
+
+    Parameters
+    ----------
+    signal : Signal
+        The linear phase filter.
+    method : str, optional
+        The method:
+            'homomorphic' (default)
+                This method works best with filters with an
+                odd number of taps, and the resulting minimum phase filter
+                will have a magnitude response that approximates the square
+                root of the the original filter's magnitude response.
+            'hilbert'
+                This method is designed to be used with equi-ripple
+                filters with unity or zero gain regions.
+    n_fft : int, optional
+        The FFT length used for calculating the cepstrum. Should be at least a
+        few times larger than the signal length. The default is ``None``,
+        resulting in an FFT length of:
+
+            n_fft = 2 ** int(np.ceil(np.log2(2 * (len(h) - 1) / 0.01)))
+
+    pad : bool, optional
+        If ``pad`` is ``True``, the resulting signal will be padded to the
+        same length as the input. The default is ``False``
+    return_magnitude_error : bool, optional
+        If ``True``, the absolute error between the input and the output is
+        returned, by default ``False``.
+
+    Returns
+    -------
+    Signal
+        The minimum phase version of the filter.
+    """
+    signal_flat = signal.flatten()
+    signal_minphase = signal.flatten()
+    signal_minphase.time = np.zeros(
+        (signal_minphase.cshape[0], int(np.floor((signal.n_samples + 1)/2))),
+        dtype=signal.dtype)
+
+    for ch in range(signal_minphase.cshape[0]):
+        signal_minphase.time[ch] = sgn.minimum_phase(
+            signal_flat.time[ch],
+            method=method,
+            n_fft=n_fft)
+
+    if (pad is True) or (return_magnitude_error is True):
+        sig_minphase_pad = pad_zeros(
+            signal_minphase, signal.n_samples - signal_minphase.n_samples)
+
+        if return_magnitude_error is False:
+            return sig_minphase_pad
+
+        error_mag = (np.abs(signal.freq) - np.abs(sig_minphase_pad.freq))
+        error = pyfar.FrequencyData(error_mag, signal.frequencies)
```

Comment:
While this is already helpful, the most important question for me when using minimum phase generation always was: What is the largest error in the region where the original signal has relatively much energy. I used the first 60 dB of the dynamic range in AKtools and returned the error in dB. Something like this, maybe:
```
error_range = np.max(np.abs(signal.freq)) * 10**(-60/20)
error_dB = np.abs(20 * np.log10(signal.freq / sig_minphase.freq))
error_dB = np.max(error_dB[np.abs(signal.freq) >= error_range])
```

Reply:
Mh, the thing is: This is a rather arbitrary measure which is likely to explode in the stop band region of band-pass filters.